### PR TITLE
feat: (observability) trace Database.batchCreateSessions + SessionPool.createSessions

### LINKED ISSUE #2146: observability: if an async context manager has not been enabled, make sure to register one
Reported offline by @alkatrivedi, whereby their code didn't involve creating a new traces at all but was using this instrumented package and their code used async/await yet spans that are supposed to be nested were disjoint.

OpenTelemetry-JS acknowledges that for code to work correctly and for context to be correctly propagated (as a replacement for threadlocal storage) a context manager must be enabled beforehand https://opentelemetry.io/docs/languages/js/context/

Customers should never have to think about these reasons and do it themselves, instead our package should automatically handle this scenario so that customers can transparently keep their code without any needs to edit
 
## Task
- [ ] Check for if a prior context manager was enabled and if not, create oen like this ideally in src/instrument.ts
```javascript
const {AsyncHooksContextManager} = require('@opentelemetry/context-async-hooks');
const {context} = require('@opentelemetry/api');
const contextManager = new AsyncHooksContextManager();
contextManager.enable();
context.setGlobalContextManager(contextManager);
```

along with regression tests for it.

### PR DESCRIPTION
**Pull request by odeke-em**
This change adds tracing for Database.batchCreateSessions
as well as SessionPool.createSessions which was raised
as a big need.
This change is a premise to finishing up tracing Transaction.

While here, also folded in the async/await fix to avoid day+ long
code review lag and then 3+ hours just to run tests per PR:
OpenTelemetry cannot work correctly for async/await if there isn't
a set AsyncHooksManager, but we should not burden our customers with
this type of specialist knowledge, their code should just work and
this change performs such a check. Later on we shall file a feature
request with the OpenTelemetry-JS API group to give us a hook to detect
if we've got a live asyncHooksManager instead of this mandatory
comparison to ROOT_CONTEXT each time.

Fixes #2146
Updates #2079
Spun out of PR #2122
Supersedes PR #2147

### CODE REVIEW

File: src/session-pool.ts
Comment (surbhigarg92) [inline src/session-pool.ts:172]: We should not add `observabilityOptions` as part of `SessionPoolOptions` . Customer pass the SessionPoolOptions and it will be confusing for them on what to use. Create this as a private property in sessionpool class, similar to _traces property

**Comment by surbhigarg92** [comment]
Can you also add record session leaks in opentelemetry traces. 

Add the [this._getLeaks()](https://github.com/googleapis/nodejs-spanner/blob/main/src/session-pool.ts#L904) in traces, also

File: src/session-pool.ts
Comment (odeke-em) [reply src/session-pool.ts]: Alright removed, thanks @surbhigarg92!

**Comment by odeke-em** [comment]
And rebased from main, good to get to bots testing and review updates @alkatrivedi @surbhigarg92!

File: observability-test/spanner.ts
Comment (alkatrivedi) [inline observability-test/spanner.ts:511]: remove ?

File: src/instrument.ts
Comment (alkatrivedi) [inline src/instrument.ts:115]: remove

File: src/instrument.ts
Comment (alkatrivedi) [inline src/instrument.ts:113]: // If no active context was set previously, trace context propagation cannot
// function correctly with async/await for OpenTelemetry
    // See {@link https://opentelemetry.io/docs/languages/js/context/#active-context}

File: src/instrument.ts
Comment (alkatrivedi) [inline src/instrument.ts:116]: remove

File: src/instrument.ts
Comment (alkatrivedi) [inline src/instrument.ts:113]:
```suggestion
    // If no active context was set previously, trace context propagation cannot
    // function correctly with async/await for OpenTelemetry
    // See {@link https://opentelemetry.io/docs/languages/js/context/#active-context}
```

File: src/instrument.ts
Comment (alkatrivedi) [inline src/instrument.ts:116]:
```suggestion
```

File: src/instrument.ts
Comment (alkatrivedi) [inline src/instrument.ts:117]:
```suggestion
    context.disable(); // Disable any prior contextManager.
```

File: observability-test/database.ts
Comment (alkatrivedi) [inline observability-test/database.ts:379]:
```suggestion
    it('should execute batchCreateSessions without error', done => {
```

File: observability-test/database.ts
Comment (alkatrivedi) [inline observability-test/database.ts:433]:
```suggestion
    it('should execute batchCreateSessions with error', done => {
```

File: observability-test/database.ts
Comment (alkatrivedi) [inline observability-test/database.ts:474]:
```suggestion
        // Ensure events are not present
```

File: observability-test/database.ts
Comment (alkatrivedi) [inline observability-test/database.ts:439]: let's not pass the session count directly, better if we make a variable first and then pass that variable to batchCreateSessions

File: observability-test/database.ts
Comment (alkatrivedi) [inline observability-test/database.ts:385]: let's not pass the session count directly, better if we make a variable first and then pass that variable to batchCreateSessions

File: src/database.ts
Comment (surbhigarg92) [inline src/database.ts:685]: Modify the variable name from `q` to `traceConfig` as discussed in previous PR

File: src/session-pool.ts
Comment (surbhigarg92) [inline src/session-pool.ts:494]: Why do we need this if check. We should simply override the observability option ?

File: src/session-pool.ts
Comment (surbhigarg92) [inline src/session-pool.ts:761]: Why do we need to set. the observability option here again ? We should be doing it in constructor ?

File: src/session-pool.ts
Comment (surbhigarg92) [inline src/session-pool.ts:792]: There should be a `span.end()` in the catch block ? before throwing error ? 
If this is true , also add a test case which could have caught this

File: observability-test/spanner.ts
Comment (surbhigarg92) [inline observability-test/spanner.ts:992]: Is this `after` in the test method ? why ?

File: observability-test/spanner.ts
Comment (surbhigarg92) [inline observability-test/spanner.ts:973]: use a better name to describe the test block. Maybe something like `E2E traces with async/await` ?

File: observability-test/spanner.ts
Comment (surbhigarg92) [inline observability-test/spanner.ts:974]: Can we also add a test case using "callback"

File: observability-test/database.ts
Comment (odeke-em) [reply observability-test/database.ts:379]: This feels pedantic/verbose and repeating also out of context because I am really testing what happens without an error returned and also with an error returned.

File: src/session-pool.ts
Comment (odeke-em) [reply src/session-pool.ts:793]: Good catch and thank you! I've added a check and a fix too.

File: observability-test/spanner.ts
Comment (odeke-em) [reply observability-test/spanner.ts:1013]: But they are really regression tests, wouldn't E2E be even more obscure? Regression to catch any future bugs given an exact use case per bug.

File: observability-test/spanner.ts
Comment (odeke-em) [reply observability-test/spanner.ts]: Yes please, and done, thank you @surbhigarg92, please take a look!

File: src/session-pool.ts
Comment (odeke-em) [reply src/session-pool.ts:494]: Not really, I was being cautious in case someone had passed in multiple observability configurations one for the explicitly created SessionPool and then one for the Database separately. It doesn't hurt to be conservative at the beginning before these libraries have been put into heavy usage.

File: src/session-pool.ts
Comment (odeke-em) [reply src/session-pool.ts:761]: There was some test case in which the observability options for database weren't set despite having been passed into the Spanner constructor.

File: observability-test/spanner.ts
Comment (odeke-em) [reply observability-test/spanner.ts]: No longer relevant but really you are allowed to place an after anywhere, especially to capture nested scope variables.

File: src/database.ts
Comment (odeke-em) [reply src/database.ts]: Indeed, thank you!

File: src/instrument.ts
Comment (odeke-em) [reply src/instrument.ts]: Done, thank you @alkatrivedi!

File: src/instrument.ts
Comment (odeke-em) [reply src/instrument.ts]: Done, thank you @alkatrivedi!

File: observability-test/spanner.ts
Comment (odeke-em) [reply observability-test/spanner.ts]: Done, thank you @alkatrivedi!

File: observability-test/database.ts
Comment (odeke-em) [reply observability-test/database.ts:439]: I don't agree, this is an inlined and singly used value, creating a variable for such a case seems unnecessary and pedantic.

File: observability-test/database.ts
Comment (odeke-em) [reply observability-test/database.ts:379]: We aren't testing Cloud Spanner's code but rather the instrumentation added and what happens with and without errors.

File: observability-test/database.ts
Comment (odeke-em) [reply observability-test/database.ts:385]: Ditto about that being pedantic yet this is a single usage.

File: observability-test/spanner.ts
Comment (odeke-em) [reply observability-test/spanner.ts]: Also btw within this section I've included the callback variant of that bug reproducer

File: observability-test/spanner.ts
Comment (surbhigarg92) [inline observability-test/spanner.ts:1013]: We can not have test block name as 'Regression tests for fixed bugs' . Block should tell what the scope of testing

File: src/instrument.ts
Comment (surbhigarg92) [reply src/instrument.ts:116]: I think we should remove this code. As we have added the opentelemetry link which should be enough. The bug will be anyway linked so maintainers would be able to track it using the code blame if required.

File: src/session-pool.ts
Comment (surbhigarg92) [reply src/session-pool.ts:494]: But we don't allow customers to pass multiple observability configurations right ? They should be able to set it only once via spannerOptions. 

We also modified the _observabilityOptions object to add _ to represent private variable.

File: src/session-pool.ts
Comment (surbhigarg92) [reply src/session-pool.ts:761]: Then we should find a way to fix those test cases and not add the code here again.

File: observability-test/spanner.ts
Comment (odeke-em) [reply observability-test/spanner.ts:1013]: I think I'll defer to you then for the naming as you are the maintainer; this because for other projects we usually have explicit sections for regression tests where it is highly encouraged that you post the reproducers. What name would you like for this?

File: observability-test/session-pool.ts
Comment (surbhigarg92) [inline observability-test/session-pool.ts:30]: Why don't we add `@opentelemetry/sdk-trace-base` as a test dependency and remove this lint from everywhere in code

**Review by surbhigarg92 — APPROVED** [review]

File: observability-test/session-pool.ts
Comment (odeke-em) [reply observability-test/session-pool.ts:30]: @surbhigarg92 it is added in `devDependencies`

Comment (surbhigarg92) [reply observability-test/session-pool.ts:30]: then why do we need to add this eslint suppress ?

Comment (odeke-em) [reply observability-test/session-pool.ts:30]: Without it we were getting build failures, it's basically the template for the tests in observability-test from long when we started sending in code changes. I really didn't dig in as it required so much time going through and figuring the build system's problems and it wasn't worth the hours sweating on.

**Pull request merged**
